### PR TITLE
init: servo at 0-unstable-2024-09-09

### LINKED ISSUE #331459: Package request: servo
**Project description**

**Metadata**

* homepage URL: https://servo.org
* source URL: https://github.com/servo/servo/
* license: MPL-2.0
* platforms: linux (not sure about other platforms due to the x11 and udev dependencies)

Depends on #331429 and https://github.com/servo/servo/pull/32896

I have already tried to package servo myself here: https://github.com/tinted-software/nix-packages/blob/main/servo.nix
However, I ran into an issue with the predefined.rs file here: https://github.com/servo/stylo/blob/f7d72a956de1e2e7455a03964af72fbbec4a9158/atoms/build.rs#L25
The atoms build.rs file tries to include a file which doesn't exist at that path. With nix it seems to be exist under `../style-0.0.1` instead of `../style` 🤔 

I would be willing to make a PR once Rust 1.80 is supported in nixpkgs, but I have no idea how to fix the above issue atm. Additionally, servo does not have any releases or tags for some reason. So I'd assume that the best way to package it would be to follow the llvmPackages_git approach: `servo-{date}"

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by supinie**
## Description of changes

a web rendering engine written in Rust, with WebGL and WebGPU support, and adaptable to desktop, mobile, and embedded applications. 

https://servo.org/
https://github.com/servo/servo
https://book.servo.org/

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/se/servo/package.nix [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+  zlib,
+  vulkan-loader,
+  wayland,
+  libGL,
+  gnumake42,
+}:
+
+rustPlatform.buildRustPackage rec {
+  pname = "servo";
+  version = "unstable-2024-08-09";
```
Comment (dotlambda) [inline pkgs/by-name/se/servo/package.nix:35]: Please read https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-naming. The commit message also has to be fixed.

File: pkgs/by-name/se/servo/package.nix [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+  src = fetchFromGitHub {
+    owner = "servo";
+    repo = pname;
```
Comment (dotlambda) [inline pkgs/by-name/se/servo/package.nix:39]:
```suggestion
    repo = "servo";
```
#277994

File: pkgs/by-name/se/servo/package.nix (hunk 1 above)
Comment (supinie) [reply pkgs/by-name/se/servo/package.nix:35]: This is still a draft and will be cleaned up before submitted. Opened as a draft for tracking purposes only at this stage.

**Comment by theoparis** [comment]
Why does this need a very old llvm version instead of using llvmPackages (which defaults to 18)? I was able to build servo without that requiremenf via cargo.

**Comment by supinie** [comment]
> Why does this need a very old llvm version instead of using llvmPackages (which defaults to 18)? I was able to build servo without that requirement via cargo.

The dependency on llvm_14 was explicitly added [here](https://github.com/servo/servo/pull/31055) to support NixOS 23.11 - Once the build is in a working state we will look at upgrading any older dependencies copied from servo's own `shell.nix` but until that point it is most likely best to keep consistency between our build and a know working state.

**Comment by honnip** [comment]
Result of `nixpkgs-review pr 337781` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>:x: 1 package failed to build:</summary>
  <ul>
    <li>servo</li>
  </ul>
</details>

---

```
error: hash mismatch in fixed-output derivation '/nix/store/9djqjfpw3zrkr4s8wqqmmlah04n8wmfz-source.drv':
        likely URL: https://github.com/servo/servo/archive/938fd8c12fc2489303e12538d3e3585bd771141f.tar.gz
         specified: sha256-Z/LQC73v3It/yMFPbUN/PPW/mLYNoGKbAoDMTMKlfyA=
            got:    sha256-CrpEBFYd8Qd0rxSnT81IvtxxEuYG0jWGJeHISvxalyY=
```

**Comment by supinie** [comment]
@honnip thank you for bringing that to my attention, there must've been something cached. I have updated all hashes to ensure they are accurate.

**Comment by ToasterUwU** [comment]
Result of `nixpkgs-review pr 337781` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>servo</li>
  </ul>
</details>

**Review by ToasterUwU — APPROVED** [review]

**Review by eljamm — COMMENTED** [review]
Thank you for your great efforts @supinie. I've tested the package for a bit and everything seems to be working well. I just have a few suggestions which I hope you can address.

**Note:** I think we should create a wrapper derivation for servo in the future so that we don't have to always re-compile it when the build doesn't change, for example if we only modify the runtime deps or env vars.

**Comment by supinie** [comment]
Hi @eljamm, thank you for your comments, they're really helpful and I will work through implementing them now. 

I'm unfamiliar with wrapper derivations, do you happen to have any resources or examples to hand? On the other hand, servo is an incredibly active repo; if we're updating deps etc. would it not make sense to bump the rev at the same time anyway which would minimise the impact that a wrapper derivation would have? As I said before though, I'm unfamiliar with this concept so more than happy to discuss more :)

**Comment by eljamm** [comment]
Firefox-based browsers use [wrapFirefox](https://github.com/NixOS/nixpkgs/blob/bf038bd724982de67a2ee9a6b2f1fcb7d6a87bd7/pkgs/applications/networking/browsers/firefox/wrapper.nix) to provide dependencies which don't need to be built with the package among other stuff. We don't need to update the main package for this to happen. If these dependencies were part of the main derivation, any update to them would lead to a full rebuild of Firefox, which is pretty resource-intensive.

Any change to other things will also cause a rebuild. For example, when we need to add an env variable or change runtime options/flags to work around an issue or enable/disable a feature.

It's probably not gonna be an issue at the moment, but it's good to keep this in mind.

Other examples:

- [kakoune](https://github.com/NixOS/nixpkgs/blob/4b99e902fb8f58d27a4c41cce94c4ad89aaced9b/pkgs/applications/editors/kakoune/wrapper.nix)
- [libreoffice](https://github.com/NixOS/nixpkgs/blob/bf038bd724982de67a2ee9a6b2f1fcb7d6a87bd7/pkgs/applications/office/libreoffice/wrapper.nix)
- [mpv](https://github.com/NixOS/nixpkgs/blob/bf038bd724982de67a2ee9a6b2f1fcb7d6a87bd7/pkgs/applications/video/mpv/wrapper.nix)
- [yazi](https://github.com/NixOS/nixpkgs/blob/bf038bd724982de67a2ee9a6b2f1fcb7d6a87bd7/pkgs/by-name/ya/yazi/package.nix)

File: pkgs/by-name/se/servo/package.nix [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+          xorg.libXi
+          libxkbcommon
+          vulkan-loader
+          wayland
+          libGL
+        ]
+      }
+  '';
+
+  # these need to be linked twice otherwise nothing will render(?!)
```
[the remark this answers is not in the document]
Comment (supinie) [reply pkgs/by-name/se/servo/package.nix:141]: You are right, before I had troubles with a white screen on servo.org but it is working now

**Review by eljamm — APPROVED** [review]
Looks good to me, thanks!

**Comment by ToasterUwU** [comment]
Result of `nixpkgs-review pr 337781` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>servo</li>
  </ul>
</details>

**Review by ToasterUwU — APPROVED** [review]

File: pkgs/by-name/se/servo/package.nix [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+  perl,
+  pkg-config,
+  python3,
+  vulkan-loader,
+  which,
+  yasm,
+  zlib,
+
+  # runtime deps
+  cargo-deny,
```
Comment (honnip) [inline pkgs/by-name/se/servo/package.nix:27]: I don't think `cargo-deny` is a runtime dependency, maybe `taplo` too?

**Review by honnip — COMMENTED** [review]
minor things

**Review by honnip — APPROVED** [review]

**Pull request merged**